### PR TITLE
#161593689 Create an album

### PR DESCRIPTION
**Pull request by JoyyToo**
#### What does this PR do?
- Add functionality to create an album

#### Description of Task to be completed?
- Create album model
- Create a user group model
- Add create album controller action
- Write tests

#### How should this be manually tested?
- Open postman and go to `localhost:3000/albums`
- Add required fields and check the response

#### What are the relevant pivotal tracker stories?
- [#161593689](https://www.pivotaltracker.com/story/show/161593689)

#### Screenshots
- Creating album with an existing user group
screen shot 2018-12-17 at 15 34 39

- Creating an album and user group at the same time
screen shot 2018-12-17 at 15 32 16

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -10,9 +10,23 @@
...
 
+  create_table "albums", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
+    t.string "title"
+    t.text "description"
+    t.jsonb "meta_data"
+    t.jsonb "user_groups"
+    t.integer "position"
+    t.bigint "tag_id"
+    t.jsonb "preferences"
+    t.integer "user_id"
```
Comment (chalchuck) [inline db/schema.rb]: All foreign+primary keys should be in either BIGINT or UUID, please refer to the ERD, hereby attached. 
image

File: config/routes.rb [hunk 2]
```diff
@@ -1,2 +1,3 @@
 Rails.application.routes.draw do
+  post 'album', to:'album#create'
```
Comment (chalchuck) [inline config/routes.rb]: use `resources :albums` to generate your routes, it more efficient. {If you only need to use `create` use the `only` option.}

`resources :people, only: [:create]`

File: app/models/album.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+class Album < ApplicationRecord
+  validates_presence_of :tag_id, :user_id, :position, :title
```
Comment (chalchuck) [inline app/models/album.rb]: - please use the shorthand validation format, it's the preferred way since rails 4.x or something like that.
- have the validations independent and configure for custom messages as shown below

`validates :tag_id, presence: true`
`validates :tag_id, presence: {message: 'tag id can not be blank for an album'}`


please arrange the model as shown in the attachment below (you may follow [this](https://github.com/rubocop-hq/rails-style-guide#activerecord) link for more  )

image

File: db/schema.rb [hunk 4]
```diff
@@ -10,9 +10,23 @@
...
 
-ActiveRecord::Schema.define(version: 0) do
+ActiveRecord::Schema.define(version: 2018_11_29_113549) do
 
   # These are extensions that must be enabled in order to support this database
+  enable_extension "pgcrypto"
   enable_extension "plpgsql"
 
+  create_table "albums", id: :uuid, default: -> { "gen_random_uuid()" }, force: :cascade do |t|
+    t.string "title"
```
[the remark this answers is not in the document]
Comment (JoyyToo) [reply db/schema.rb]: Hey @chalchuck do I have to add constraints in the db when they are already handled in the model? For example I have a validate presence title of in the model should i also add a `null: false` in the db?

File: db/schema.rb [hunk 5]
```diff
@@ -10,9 +10,23 @@
...
+    t.text "description"
+    t.jsonb "meta_data"
+    t.jsonb "user_groups"
+    t.integer "position"
```
[the remark this answers is not in the document]
Comment (JoyyToo) [reply db/schema.rb]: So in this case, if we have a default and we are also incrementing it, and the erd states that it is should be a mandatory field, it will be conflicting as the user we also be asked to input the position. So do I remove it as a mandatory field?

[the remark this answers is not in the document]
Comment (chalchuck) [reply db/schema.rb]: yes please, remove it as a mandatory field. @gr1d99 please make this change on the ERD

For the user editing the position; this will be made possible using the **Drag and Drop** feature from the front end; on the back end this will be handled/achieved by [Acts as List](https://github.com/swanandp/acts_as_list)

File: app/controllers/album_controller.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    if @album.save
+      render json: @album, status: :created
+    end
+  end
+
+  private
+
+  def album_params
+    params.permit(:title, :description, :meta_data, :user_groups, :position,
+                  :tag_id, :user_id, :preferences)
```
Comment (gr1d99) [inline app/controllers/album_controller.rb]: make sure you wrap album parameters with `album` as it key so that you can make use of `permit`,
  eg 
```ruby
params.require(:album).permit({...})
```

File: spec/controllers/album_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+require "rails_helper"
+
+RSpec.describe AlbumController, type: :controller do
+  let!(:album) { create(:album) }
+  let(:album_params) { attributes_for(:album) }
+
+  describe "POST /album" do
```
Comment (gr1d99) [inline spec/controllers/album_controller_spec.rb]: make sure you add tests for failures

File: app/controllers/album_controller.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class AlbumController < ApplicationController
+  def new
```
[the remark this answers is not in the document]
Comment (chalchuck) [reply app/controllers/album_controller.rb]: When building API only rails apps, we only have `%w(index show create update destroy) `actions only.

File: spec/controllers/album_controller_spec.rb (hunk 7 above)
Comment (chalchuck) [reply spec/controllers/album_controller_spec.rb]: When doing tests, it's always a good practice to always have the tests covering ALL possible edge cases; including but not limited to

- Bad user input (submit 100 chars instead of the required 64/32)
- Invalid user input (submit an invalid album title, for example a title containing special chars .. !@#$%^&*)
- Please remember to validate the album title to only take alphanumerics

**Comment by chalchuck** [comment]
@JoyyToo When displaying JSON data, it's recommended to conform to the [jsonapi](https://jsonapi.org/) Spec. 

To achieve this in Ruby on Rails, we use POROS (Plain Ruby Objects) called Serializers, these serializers are provided by some third part gems, some of them are:

1. [jsonapi-rb](http://jsonapi-rb.org/) is a highly performant and modular JSON:API-only implementation. There's a vibrant community around it that has produced projects such as JSON:API Suite.
2. [fast_jsonapi](https://github.com/Netflix/fast_jsonapi) is a lightning fast JSON:API serializer for Ruby Objects from the team of Netflix.

But for the sake of simplicity, we are going to use [Active Model Serializers (AMS)](https://github.com/rails-api/active_model_serializers) for this project.

Once configured, the JSON output will have to have the following structure, and it has to include `type, id and attributes` fields

```
 {
    "type": "comments",
    "id": "5",
    "attributes": {
      "body": "First!"
   }
}
```

have a look at the herein attached LIVE example of such an output

image


Please reach out to me in person/slack if you face any issues or need any clarification.

**Comment by JoyyToo** [comment]
@chalchuck @gr1d99 @Orelongz I have implemented the feedback above

File: spec/models/album_spec.rb [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+        with_message("Cannot be blank for an album")
+    }
+    it {
+      should validate_presence_of(:tag_id).
+        with_message("Cannot be blank for an album")
+    }
+    it {
+      should validate_presence_of(:user_id).
+        with_message("Cannot be blank for an album")
+    }
```
Comment (kasigazibrian) [inline spec/models/album_spec.rb]: Great work @JoyyToo .  I noticed you have validation for the format of the title. you could use this format to test for its format. Something like this.
```
 it do
    should allow_value('valid-title').for(:title)
 end
```
```
  it do
     should_not allow_value('invalid-title').for(:title)
  end
```
You could also add a test for the length using this format
```
 it do
     should validate_length_of(:title).is_at_least(64)
 end
```
You can find more info on this [link](https://makandracards.com/makandra/643-testing-validates_format_of-with-shoulda-matchers)

Comment (JoyyToo) [reply spec/models/album_spec.rb]: In my controller tests, I have already captured for these bad user inputs

File: app/controllers/album_controller.rb [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  private
+
+  def album_params
+    params.require(:album).permit(:title, :description, :user_id, :position,
+                                  :metadata, :preferences, tag_list: [])
+  end
+
+  def user_groups_params
```
Comment (chalchuck) [inline app/controllers/album_controller.rb]: you do not need to have this here, extract it to the `CreateAlbumService`

File: app/controllers/album_controller.rb [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    params.fetch(:user_groups, {}).permit(:name, emails: [])
+  end
+
+  def user_groups_ids
```
Comment (chalchuck) [inline app/controllers/album_controller.rb]: you do not need to have this here, extract it to the `CreateAlbumService`

File: app/forms/album_form.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+                    length: { maximum: 64 }, format: {
+                      with: TITLE_REGEXP,
+                      message: I18n.t("message.album.alphanumerics"),
+                    },
+                    presence: { message: I18n.t("message.album.blank") }
+
+  validates :user_id, presence: { message: I18n.t("message.album.blank") }
+
+  collection :user_groups, populate_if_empty: UserGroup do
+    properties :name, :emails
```
Comment (chalchuck) [inline app/forms/album_form.rb]: Add a default value of {} for emails, right here in the `FormObject`

File: app/models/album.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+                    uniqueness: { message: I18n.t("message.album.unique") },
+                    length: { maximum: 64 }, format: {
+                      with: TITLE_REGEXP,
+                      message: I18n.t("message.album.alphanumerics"),
+                    }
+
+  # callbacks
+  before_create :set_position
+
+  def set_position
```
Comment (chalchuck) [inline app/models/album.rb]: I would prefer a PostgreSQL TRIGGER function for this.
Did you try `album.increment!(:position)`

File: app/services/albums/create_album_service.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
+module Albums
+  class CreateAlbumService
+    def initialize(album_params, user_group_params, user_group_ids)
```
Comment (chalchuck) [inline app/services/albums/create_album_service.rb]: you only need the `params` the rest can be derived from from params

Comment (chalchuck) [inline app/services/albums/create_album_service.rb]: Let us make the method call be `Albums:: CreateAlbumService .call(params)`

File: app/services/albums/create_album_service.rb [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+      @album_params = album_params
+      @user_group_params = user_group_params
+      @user_group_ids = user_group_ids
+    end
+
+    def call
```
Comment (chalchuck) [inline app/services/albums/create_album_service.rb]: this method is very long. `call` should only invoke methods that perform the task at hand

File: app/services/albums/create_album_service.rb [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+      form = AlbumForm.new(Album.new)
+      unless user_group_params.empty?
+        album_params[:user_groups] = [user_group_params]
+      end
+      form.validate(album_params)
```
Comment (chalchuck) [inline app/services/albums/create_album_service.rb]: this is incorrect, 
it should be something like this 

```
if form.validate(album_params)
    form.save
....
....
.......
```

File: spec/services/create_album_spec.rb [hunk 17]
```diff
@@ -0,0 +1,53 @@
+describe Albums::CreateAlbumService do
+  before do
+    @service = Albums::CreateAlbumService
```
Comment (gr1d99) [inline spec/services/create_album_spec.rb]: use `Use let and let!` to hold your variables

File: spec/services/create_album_spec.rb [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+        "tag_list": %w(thanksgiving halloween),
+      },
+      "user_groups": {
+        "name": "Test",
+        "emails": %w(test@gmail.com user@gmail.com),
+      },
+    }
+  end
+
+  context "when the request is valid " do
```
Comment (gr1d99) [inline spec/services/create_album_spec.rb]: `context "when the request is valid "` is incorrect context description, we are not making any request to this service object, instead, we are passing params as arguments
